### PR TITLE
feat: allow <mux-video> and <mux-player> to get any metadata-* attrs set

### PR DESCRIPTION
**Pull request by dylanjha**
Related to #500 

Keeping this as a draft rn to answer this question:

1. Is there any value to keeping the well-defined list of 3 "special" attributes? `Attributes.METADATA_VIDEO_ID`, etc? Autocomplete tooling, typescript flagging in editors? 
2. For `mux-player`, does it makes sense to use the `connectedCallback` in order to access the shadow DOMs `muxVideo.metadata=` prop?

# todo

- [x] resolve outstanding questions
- [x] tests

# mux-video

This solves it for `<mux-video>` so someone could do:

```html
<mux-video
  <!-- other attrs -->
  metadata-sub-property-id="2434"
></mux-video>
```

sub-id_2022-11-17_13-26-21
sub-id_2022-11-17_13-26-09

- [x] Also we'd need to do this for `<mux-audio>` (same code as this).

# mux-player

```html
<mux-player
  <!-- other attrs -->
  metadata-sub-property-id="2434"
></mux-player>
```

Makes a small internal change to how metadata is passed from `<mux-player>` -> `<mux-video>`. Now we *only* use the `muxVideo.metadata=` prop for assigning metadata. The attrs syntax is removed from the template

### CODE REVIEW

File: packages/mux-video/src/index.ts [hunk 1]
```diff
@@ -377,11 +377,30 @@ class MuxVideoElement extends CustomVideoElement<HTMLVideoElement> implements Pa
...
     const viewer_user_id = this.getAttribute(Attributes.METADATA_VIEWER_USER_ID);
+
+    const otherMetadataAttrs: { [key: string]: string } = {};
+    this.getAttributeNames()
+      .filter((attrName) => {
+        return (
+          attrName.startsWith('metadata-') &&
+          ![Attributes.METADATA_VIDEO_ID, Attributes.METADATA_VIDEO_TITLE, Attributes.METADATA_VIEWER_USER_ID].includes(
+            attrName
+          )
```
Comment (dylanjha) [inline packages/mux-video/src/index.ts:388]: TS error:

```
src/index.ts:387:13 - error TS2345: Argument of type 'string' is not assignable to parameter of type '"metadata-video-id" | "metadata-video-title" | "metadata-viewer-user-id"'.

387             attrName
                ~~~~~~~~
```

File: packages/mux-player/src/index.ts [hunk 2]
```diff
@@ -237,8 +237,25 @@ class MuxPlayerElement extends VideoApiElement {
...
+        }
+      });
+    return metadata;
+  }
+
   connectedCallback() {
     this.#renderChrome();
+    const muxVideo = this.shadowRoot?.querySelector('mux-video') as MuxVideoElement;
+    if (muxVideo) {
+      muxVideo.metadata = metadataFromAttrs();
```
Comment (cjpillsbury) [inline packages/mux-player/src/index.ts:257]: issue(blocking): `muxVideo.metadata = this.metadataFromAttrs`

File: packages/mux-player/src/index.ts [hunk 3]
```diff
@@ -1084,7 +1101,7 @@ class MuxPlayerElement extends VideoApiElement {
       logger.error('underlying media element missing when trying to set metadata. metadata will not be set.');
       return;
     }
-    this.media.metadata = val;
+    this.media.metadata = { ...metadataFromAttrs, ...val };
```
Comment (cjpillsbury) [inline packages/mux-player/src/index.ts:1104]: issue(blocking): `this.media.metadata = { ...this.metadataFromAttrs, ...val }`

File: packages/mux-video/src/index.ts [hunk 4]
```diff
@@ -362,11 +362,34 @@ class MuxVideoElement extends CustomVideoElement<HTMLVideoElement> implements Pa
     const video_id = this.getAttribute(Attributes.METADATA_VIDEO_ID);
     const video_title = this.getAttribute(Attributes.METADATA_VIDEO_TITLE);
     const viewer_user_id = this.getAttribute(Attributes.METADATA_VIEWER_USER_ID);
+
+    const otherMetadataAttrs: { [key: string]: string } = {};
```
Comment (cjpillsbury) [inline packages/mux-video/src/index.ts:366]: suggestion(non-blocking): Per your callout in your PR description, we shouldn't need to treat the "well-defined metadata attributes" differently here. I'd remove them so we can simplify this code.

**Review by cjpillsbury — CHANGES_REQUESTED** [review]
a couple of quick fixes but otherwise 👌

**Comment by dylanjha** [comment]
Update:

- going to try to add some tests
- examples have been updated so we can test the functionality on the deploy preview -- will revert those example updates before merging
- will want to squash & merge this one with a conventional commit message like:

> feat: allow arbitrary Mux Data `metadata-*` attributes syntax (previously only supported this syntax for `metadata-video-title`, `metadata-video-id` and `metadata-viewer-user-id`

File: packages/mux-audio/src/index.ts [hunk 5]
```diff
@@ -254,7 +248,22 @@ class MuxAudioElement extends CustomAudioElement<HTMLAudioElement> implements Pa
   }
 
   get metadata() {
-    return this.#metadata;
+    const inferredMetadataAttrs: { [key: string]: string } = {};
+    this.getAttributeNames()
+      .filter((attrName) => {
+        return attrName.startsWith('metadata-') && !([Attributes.METADATA_URL] as string[]).includes(attrName);
+      })
+      .map((attrName) => {
```
Comment (cjpillsbury) [inline packages/mux-audio/src/index.ts:256]: nitpick(blocking): either use `forEach` if you're producing a side effect (like your implementation in `mux-player`) or refactor to use `reduce()`, e.g.:

```ts
const inferredMetadataAttrs: { [key: string]: string } = this.getAttributeNames()
      .filter((attrName) => {
        return attrName.startsWith('metadata-') && !([Attributes.METADATA_URL] as string[]).includes(attrName);
      })
      .reduce((curAttrs, attrName) => {
        const value = this.getAttribute(attrName);
        if (value != null) {
          curAttrs[attrName.replace(/^metadata-/, '').replace(/-/g, '_') as string] = value;
        }
        return currAttrs;
      }, {});
```

File: packages/mux-audio/test/player.test.js [hunk 6]
```diff
@@ -91,4 +91,15 @@ describe('<mux-audio>', () => {
     await aTimeout(3000);
     assert.equal(player.buffered.length, 0, 'no buffer loaded');
   });
+
+  it('maps arbitrary metadata-* attrs to the metadata prop and populates video_id if not provided', async function () {
```
Comment (cjpillsbury) [inline packages/mux-audio/test/player.test.js:95]: 😎

File: packages/mux-video/src/index.ts [hunk 7]
```diff
@@ -361,14 +355,21 @@ class MuxVideoElement extends CustomVideoElement<HTMLVideoElement> implements Pa
...
   get metadata() {
-    const video_id = this.getAttribute(Attributes.METADATA_VIDEO_ID);
-    const video_title = this.getAttribute(Attributes.METADATA_VIDEO_TITLE);
-    const viewer_user_id = this.getAttribute(Attributes.METADATA_VIEWER_USER_ID);
+    const inferredMetadataAttrs: { [key: string]: string } = {};
+    this.getAttributeNames()
+      .filter((attrName) => {
+        return attrName.startsWith('metadata-') && !([Attributes.METADATA_URL] as string[]).includes(attrName);
+      })
+      .map((attrName) => {
```
Comment (cjpillsbury) [inline packages/mux-video/src/index.ts:363]: nitpick(blocking): same feedback as `mux-audio`

**Review by cjpillsbury — CHANGES_REQUESTED** [review]
A couple of small asks. We also can end up in a weird state if folks set `metadata-*` attributes and the `metadata` object after initial load that wouldn't be the case with the current impl for our "well-defined" metadata attrs that we may want to solve.

File: packages/mux-audio/src/index.ts (hunk 5 above)
Comment (dylanjha) [reply packages/mux-audio/src/index.ts]: good call!

**Review by cjpillsbury — APPROVED** [review]
🎉

**Comment by dylanjha** [comment]
@cjpillsbury if I understand correctly there may be some edge-case breaking change if a user tries to set the same metadata fields in 2 different ways:

```html
<mux-player video-title="Title AB"></mux-player>
```

AND

```js
document.querySelector('mux-player').metadata = { video_title: 'XY' }
```

The potential change in behavior would be which one gets precedence?

If that's the case, I think we can live with that. Folks really shouldn't be setting the same metadata fields with different values in 2 different ways.

**Comment by dylanjha** [comment]
Verified the behavior on esm-vanilla deploy preview, reverted the changes to `examples/`, will squash merge after the CI builds

**Comment by cjpillsbury** [comment]
> @cjpillsbury if I understand correctly there may be some edge-case breaking change if a user tries to set the same metadata fields in 2 different ways:
> 
> ```
> <mux-player video-title="Title AB"></mux-player>
> ```
> 
> AND
> 
> ```js
> document.querySelector('mux-player').metadata = { video_title: 'Title XY' }
> ```
> 
> The potential change in behavior would be which one gets precedence?
> 
> If that's the case, I think we can live with that. Folks really shouldn't be setting the same metadata fields with different values in 2 different ways.

Kind of... before, neither ended up getting precedence because of the way the metadata attrs" flowed to" the underlying `mux-video` instance + the fact that the the getter for `metadata` always merged the set prop plus the attrs. Since we now never pass the `mux-player` attributes down via attributes, we loose this "automatic merging" behavior.

**Pull request merged**